controllers/cruisecontroltask_controller.go

```diff
@@ -62,448 +59,412 @@ func (r *CruiseControlTaskReconciler) Reconcile(ctx context.Context, request ctr
-		}
+	log.V(1).Info("reconciling Cruise Control tasks")
 
-		volumesState := make(map[string]kafkav1beta1.VolumeState)
-		for mountPath, volumeState := range brokerStatus.GracefulActionState.VolumeStates {
-			if volumeState.CruiseControlVolumeState == kafkav1beta1.GracefulDiskRebalanceRunning {
-				volumesState[mountPath] = volumeState
-			}
-		}
-		if len(volumesState) > 0 {
-			brokerVolumesWithRunningCCTask[brokerId] = volumesState
-		}
+	// Get all active tasks reported in status of Kafka Cluster CR
+	tasksAndStates := getActiveTasksFromCluster(instance)
+	if tasksAndStates.IsEmpty() {
+		log.Info("no active tasks found in Kafka Cluster status")
+		return reconciled()
 	}
-	if len(brokersWithRunningCCTask) > 0 {
-		err = r.checkCCTaskState(instance, brokersWithRunningCCTask, log)
+
+	scaler, err := scale.NewCruiseControlScaler(ctx, scale.CruiseControlURLFromKafkaCluster(instance))
+	if err != nil {
+		return requeueWithError(log, "failed to create Cruise Control Scaler instance", err)
 	}
 
-	if err == nil && len(brokerVolumesWithRunningCCTask) > 0 {
-		err = r.checkVolumeCCTaskState(instance, brokerVolumesWithRunningCCTask, log)
+	if !scaler.IsUp() {
+		log.Info("requeue event as Cruise Control is not available (yet)")
+		return requeueAfter(DefaultRequeueAfterTimeInSec)
 	}
 
+	// Update task states with information from Cruise Control
+	err = updateActiveTasks(scaler, tasksAndStates)
 	if err != nil {
-		switch errors.Cause(err).(type) {
-		case errorfactory.CruiseControlNotReady, errorfactory.ResourceNotReady:
-			return ctrl.Result{
-				RequeueAfter: time.Duration(15) * time.Second,
-			}, nil
-		case errorfactory.CruiseControlTaskRunning:
-			return ctrl.Result{
-				RequeueAfter: time.Duration(20) * time.Second,
-			}, nil
-		case errorfactory.CruiseControlTaskTimeout, errorfactory.CruiseControlTaskFailure:
-			return ctrl.Result{
-				RequeueAfter: time.Duration(20) * time.Second,
-			}, nil
-		default:
-			return requeueWithError(log, err.Error(), err)
+		log.Error(err, "requeue event as updating state of active tasks failed")
+		return requeueAfter(DefaultRequeueAfterTimeInSec)
+	}
+
+	// Check if CruiseControl is ready as we cannot perform any operation until it is in ready state
+	if status := scaler.Status(); status.InExecution() {
+		log.Info("updating status of Kafka Cluster and requeue event as Cruise Control is in execution")
+		if err := r.UpdateStatus(ctx, instance, tasksAndStates); err != nil {
+			log.Error(err, "failed to update Kafka Cluster status")
 		}
+		return requeueAfter(DefaultRequeueAfterTimeInSec)
 	}
 
-	var brokersWithDownscaleRequired []string
-	var brokersWithUpscaleRequired []string
-	brokersWithDiskRebalanceRequired := make(map[string][]string)
+	switch {
+	case tasksAndStates.NumActiveTasksByOp(OperationAddBroker) > 0:
+		addBrokerTasks := make([]*CruiseControlTask, 0)
+		brokerIDs := make([]string, 0)
+		for _, task := range tasksAndStates.GetActiveTasksByOp(OperationAddBroker) {
```

Comment:
Could be there any edge case when the list of broker IDs collected from `tasksAndStates.GetActiveTasksByOp` contains duplicate which later may cause an issue when  `scaler.AddBrokers` is invoked. If that's the case `bokerIDs` should be rather a `map[string]struct{}` where the key is broker id

Reply:
There should not be any duplicated entries. Having duplicated entries indicates an issue in the controller logic which we want to surface as soon as possible. Also Cruise Control API is able to handle duplicated entries for broker IDs, so it does not trigger an error on Cruise Control end.